apps/learning-snippets/src/test/property-grid/Telemetry.test.tsx

```diff
@@ -0,0 +1,150 @@
+        const logPerformance = sinon.spy();
+        const logUsage = sinon.spy();
+        // __PUBLISH_EXTRACT_START__ PropertyGrid.ComponentWithTelemetry
+        const MyPropertyGrid = () => {
+          return (
+            <PropertyGridComponent
+              onPerformanceMeasured={(feature, elapsedTime) => {
+                // user-defined function to handle performance logging. 
+                logPerformance(feature, elapsedTime)
+              }}
+              onFeatureUsed={(feature) => {
+                // user-defined function to handle usage logging. 
+                logUsage(feature);
+              }}
+            />
+          );
+        };
+        // __PUBLISH_EXTRACT_END__
+        render(<MyPropertyGrid />);
+        Presentation.selection.addToSelection("", imodel.imodel, [imodel.category]);
+        await waitFor(() => {
+          expect(logUsage).to.be.calledTwice;
+          expect(logPerformance).to.be.calledOnce;
+        });
+        Presentation.selection.clearSelection("", imodel.imodel);
+        cleanup();
+      });
+
+      it("Renders property grid with telemetry context", async function () {
+        const imodel = (
+          await buildIModel(this, async (builder) => {
+            const physicalModel = insertPhysicalModelWithPartition({ builder, codeValue: "TestPhysicalModel" });
+            const category = insertSpatialCategory({ builder, codeValue: "Test SpatialCategory" });
+            insertPhysicalElement({ builder, modelId: physicalModel.id, categoryId: category.id });
+            return { category };
+          })
+        );
+        const imodelConnection = imodel.imodel;
+        sinon.stub(UiFramework, "getIModelConnection").returns(imodelConnection);
+        const logPerformance = sinon.spy();
+        const logUsage = sinon.spy();
+        // __PUBLISH_EXTRACT_START__ PropertyGrid.ComponentWithTelemetryWrapper
+        function ExampleContextMenuItem() {
+          return (
+            <TelemetryContextProvider
+              onPerformanceMeasured={(feature, elapsedTime) => {
+                // user-defined function to handle performance logging. 
+                logPerformance(feature, elapsedTime)
+              }}
+              onFeatureUsed={(feature) => {
+                // user-defined function to handle usage logging. 
+                logUsage(feature);
+              }}
+            >
+              <PropertyGrid imodel={imodelConnection} />
+            </TelemetryContextProvider>
+          );
+        }
+        // __PUBLISH_EXTRACT_END__
+        const user = userEvent.setup();
+        const { queryByText, getByRole } = render(<ExampleContextMenuItem />);
+        Presentation.selection.addToSelection("", imodel.imodel, [imodel.category]);
+        await waitFor(async () => {
+          expect(logPerformance).to.be.calledOnce;
+          // trigger a feature
+          const button = queryByText("search-bar.open");
```

Comment:
I would suggest using `getByRole` to find buttons or `getByText` if for some reason `getByRole` does not fit.

Reply:
Quite a weird situation with getByText and getByRole. Because the button gets pressed and the text changes after the `getByText` call it somehow manages to fail :D.
![image](https://github.com/user-attachments/assets/abb60332-76ae-4929-945a-300844124631)
![image](https://github.com/user-attachments/assets/3c4340ea-494c-4c4d-a485-0ecff4d7c912)

Reply:
I think the problem is that button lookup, click and assertion are in the same waitFor call. I think correct workflow should be 
```ts
const button = await waitFor(() => getByRole("button"));
await user.click(button);

// fill input

await waiFor(() => {
  expect(spy).to.be.called
});
```

Reply:
Changed, was under the impresion you can only have 1 waitFor.